README

		       THE UTAH RASTER TOOLKIT
                      Version 3.1c - Dec 5, 2012

The Utah Raster toolkit is a collection of programs and C routines for
dealing with raster images commonly encountered in computer graphics.  It
provides the following major functions:

	* A device and system independent image format for storing images
	  and information about them.  Called the RLE format, it uses
	  run length encoding to reduce storage space for most images.

	* A library of C routines for reading, writing and manipulating
	  images stored in the RLE format.

	* A collections of programs for manipulating and displaying RLE
	  images.

The Format:

  The device independent RLE file has two parts, the header, which stores
  information about the image (size, position, channel information,
  color maps, comments, etc), and the actual image data in a run length
  encoded format.  The RLE format often requires about a third of the
  available space for most "image synthesis" style images.  If the image
  does not compress well, the RLE format stores it as straight pixel data
  with little extra overhead.  The format was first developed around
  1983 at the University of Utah.

The Library:

  C routines are provided for setting up and reading the image header,
  and for reading and writing the image a scanline at a time.  Images can
  be read or written using two different methods.  Using the "row" method,
  the library performs the RLE encoding and decoding.  With the "raw" method,
  scanlines are constructed directly with RLE opcodes.  Additional routines
  are available for generating dither matrices (e.g., for display programs
  running on devices with less than 24 bits of color).


The Tools:
  applymap   - Apply color map values to pixel values.
  avg4 	     - Downfilter an image by 1/4, generating a matte channel if one
	       didn't previously exist
  crop 	     - Crop an image.
  dvirle     - Convert TeX output into anti-aliased images.
  fant 	     - Rotate and/or scale in image by an arbitrary (float) value.
  mcut       - Quantize an image from 24 to eight bits using the median cut
	       algorithm.
  mergechan  - Merge several channels from different files into a single
	       RLE file.
  pyrmask    - Blend images using Gaussian pyramids.
  repos      - Change the position in the RLE header.
  rleClock   - Generate an image of a clock.
  rleaddcom  - Add a comment to the RLE file's header.
  rlebg      - Generate a solid or variable background.
  rlebox     - Find the actual non-background area of an image.
  rlecomp    - Digital image compositor.  Provides the operations over, atop,
	       in, out, xor, plus, minus and diff on two images.
  rledither  - Floyd-Steinberg to a given color map.
  rleflip    - Rotate an image by 90/180 degree increments.
  rlehdr     - Dump the contents of the RLE header in human readable form.
  rlehisto   - Generate the histogram of an RLE file.
  rleldmap   - Load a color map into an RLE file from a variety of sources.
  rlemandl   - Generate Mandlebrot sets as RLE files.
  rlenoise   - Adds random noise to an image.
  rlepatch   - Overlay several smaller images over a larger one.
  rleprint   - Print all pixel values.
  rlequant   - Variance-based color quantization.
  rlescale   - Generates gray scale and color scale RLE files.
  rleselect  - Select specific images from a file.
  rlesetbg   - Set the background color stored in the RLE header.
  rleskel    - A skeleton application, for rolling your own.
  rlespiff   - Simple contrast enhancement to "spiff up" images.
  rlesplice  - Splice images together horizontally or vertically.
  rlesplit   - Split a file containing several images into several files.
  rleswap    - Swap, copy or delete channels in an RLE file.
  rlezoom    - Enlarge or shrink an image with pixel replication.
  smush      - Perform a simple Gaussian filter on an image.
  to8 	     - Convert a 24 bit RGB image to an eight bit dithered one.
  tobw 	     - Convert 24 bits to 8 bits black and white.
  unexp      - Convert an "exponential" image to a displayable one.
  unslice    - Quickly assemble an image from several horizontal strips

  Format conversion programs are provided for:
	- Ascii (line printer pictures) (to)
	- Cubicomp image format (from)
	- GIF (to and from)
	- MacPaint (to and from)
	- PBMPLUS pgm (from) and ppm (to and from)
	- PostScript (to)
	- Sun rasterfiles (to and from)
	- Targa image format (from)
	- TIFF (to and from)
	- Wasatch paint systems (from)
	- Wavefront 'RLA' format (to and from)
   	- Simple pixel streams (color & B&W) (to and from)

The display programs included with the original URT distributions have been
removed.  If converting to other formats proves insufficient for debugging
purposes, a cross-platform viewer based on a toolkit like FLTK would be more
appropriate to modern hardware than the original special purpose get tools.



ChangeLog
commit 0d78dcb56d1834ed6b49eb7729333fd18c7d1521
Author: tbrowder2 <tbrowder2@localhost>
Date:   Wed Jun 6 16:40:38 2012 +0000

    correct grammar: it's => its (possessive form of 'it')

commit 162019d267de7b2e810f33c98a4d488a8a773285
Author: Clifford Yapp <cliffyapp@gmail.com>
Date:   Wed May 2 19:49:44 2012 +0000

    Switch other src/other Makefile.am files to just straight-up ignore CMake dir

commit b946c6070077e66f69514e14a3726adeba7a16e5
Author: Clifford Yapp <cliffyapp@gmail.com>
Date:   Tue Feb 7 23:56:12 2012 +0000

    Hopefully this will get things cleaned up on Windows again.  Untested.

commit 002769220c1b4c243a5b01fdb836387b1f43debc
Author: Clifford Yapp <cliffyapp@gmail.com>
Date:   Thu Feb 2 22:00:34 2012 +0000

    Try to be better about using misc/CMake .cmake files by default.  Tweak the listing of additional files to clean from lemon output.

commit 23923e05a494f97c5e317903f1f7be1615ef4245
Author: Clifford Yapp <cliffyapp@gmail.com>
Date:   Thu Feb 2 21:02:48 2012 +0000

    Clean up some ws in src/other

commit 522e63cc8e7ce1962f5153d93085fc836269f964
Author: Clifford Yapp <cliffyapp@gmail.com>
Date:   Tue Nov 29 18:06:57 2011 +0000

    Hmm, must have had a stale libutahrle lib file around.  Tweak MSVC build flags.

commit b604d270a3ca710e1aec16432b1461786df879dc
Author: Clifford Yapp <cliffyapp@gmail.com>
Date:   Tue Nov 29 02:14:59 2011 +0000

    Make a stab at dealing with more MSVC build issues with new setup.  Not tested yet.

commit 03d2f906642819ba7161d68a05e8216b0d6d535c
Author: Christopher Sean Morrison <morrison@brlcad.org>
Date:   Fri Sep 9 21:08:06 2011 +0000

    ws consistency removing trailing ws and much more (see sh/ws.sh)

commit 156b84133e6e738db60f6bc35393b7c2e0950aaa
Author: Erik Greenwald <erikg@arl.army.mil>
Date:   Mon Jun 6 19:36:22 2011 +0000

    add CMake/* stuff to EXTRA_DIST

commit e7247705bdbb770d51ad417ff489267980a7224c
Author: Erik Greenwald <erikg@arl.army.mil>
Date:   Mon Jun 6 18:40:56 2011 +0000

    Add CMakeLists.txt to EXTRA_DIST.

commit 84e32a1ed8dc9052f8f2d7a517f0aeef8d96aafc
Author: Christopher Sean Morrison <morrison@brlcad.org>
Date:   Wed Apr 20 20:26:01 2011 +0000

    likewise with the src/other CMake dirs

commit 7067e47f4ed275ced51c0d11c3b15fc0ca33954a
Author: Christopher Sean Morrison <morrison@brlcad.org>
Date:   Wed Apr 20 18:39:37 2011 +0000

    sync all of the CMakeLists.txt files from the cmake branch to trunk, preserving history via svn cp for the new additions and copying for the rest.

commit 6179355a27744a1924528f7dbf04c7d9fa8eb539
Author: Christopher Sean Morrison <morrison@brlcad.org>
Date:   Fri Feb 18 04:37:49 2011 +0000

    check if defined before checking value

commit 0db425d1dd6db8fd4e9ce1bf8780ce0ae9447dd6
Author: Christopher Sean Morrison <morrison@brlcad.org>
Date:   Thu Apr 22 18:38:19 2010 +0000

    add the NOWARN flag everywhere effectively disabling compilation warnings for all of our external dependencies (except for libz, libpng, and libregex just because they're already fairly quiet).  clean up CFLAGS in step where they should be CPPFLAGS as well.

commit adf79b9d2259130f13074060730bd4ee1df2e070
Author: Robert Parker <bobparker@brlcad.org>
Date:   Thu Feb 4 00:46:50 2010 +0000

    Quell a few warnings when compiling for 64-bit Windows.

commit c94525c4476962f116af4364f4209c328ed03417
Author: Clifford Yapp <cliffyapp@gmail.com>
Date:   Fri Jun 5 13:33:08 2009 +0000

    Remove duplicate entry for librle.3

commit c681e8e886b94f477228ab37fee60113df0543e4
Author: Christopher Sean Morrison <morrison@brlcad.org>
Date:   Wed Nov 5 04:48:34 2008 +0000

    alrighty, it's been years of fighting merges with rcs variables getting in the way and in all that time I can't think of a time where I found that line actually incredibly useful.  time to give it the axe, at least in the makefile templates.

commit 453b033eb3f94b5a1438120a634f5c165730f4c7
Author: Christopher Sean Morrison <morrison@brlcad.org>
Date:   Tue Sep 30 08:03:46 2008 +0000

    variety of manual page fixes by Giuseppe Iuculano ( derevko ) via sf patch 2133680 (Minor fixes to manual pages)

commit 9c7cf64d2eed210e7231fc3274cc0b1c1775b04a
Author: Clifford Yapp <cliffyapp@gmail.com>
Date:   Sat Sep 20 22:58:54 2008 +0000

    make rleinterp man page use rleinterp instead of rleintrp.

commit 6b36ed3176e56fd60db37e7244164be871d457a8
Author: sammy <sammy@localhost>
Date:   Fri Jul 11 15:25:29 2008 +0000

      * rlehdr.c: fix a potential crash in the RLE header display function.

commit 128911116b94b6396824d1ed0c03cff7191c05b4
Author: Christopher Sean Morrison <morrison@brlcad.org>
Date:   Tue May 6 17:28:32 2008 +0000

    quell decl warnings

commit c3d3c32b0987cbd8e67574a85d3b897837e7105b
Author: Christopher Sean Morrison <morrison@brlcad.org>
Date:   Tue May 6 17:25:41 2008 +0000

    quell const qualifier discardation warnings

commit fc856e4f5663d5995d99bcde31239a7783a0cf6c
Author: Christopher Sean Morrison <morrison@brlcad.org>
Date:   Sun Feb 10 16:37:21 2008 +0000

    quellage

commit 3d5b06777fd66a03b33ee90e02dd217088a9decc
Author: Christopher Sean Morrison <morrison@brlcad.org>
Date:   Sun Feb 10 16:32:11 2008 +0000

    quell compilation warnings

commit b1a5cecaf9c8b286e32a513ebca61bdeff8d6328
Author: Christopher Sean Morrison <morrison@brlcad.org>
Date:   Mon Jan 14 17:08:02 2008 +0000

    use consistent spacing around the common header inclusion using one and only blank line before and after.

commit 730c41600ff437396328af4c85cbedec4a89d747
Author: Christopher Sean Morrison <morrison@brlcad.org>
Date:   Mon Jan 14 15:09:25 2008 +0000

    remove the .cvsignore files since they are no longer used.  the ignore patterns are now svn:ignore properties on the directory and can be edited with propset/propedit.

commit 816b6f6d405f6818a387b22e98505ddf8cc18099
Author: Robert Parker <bobparker@brlcad.org>
Date:   Thu Dec 20 20:36:39 2007 +0000

    Mods to get things compiling on windows.


include/rle_config.h
/** @addtogroup utahrle */
/** @{ */

/** @file rle_config.h
 *
 * Automatically generated by make-config-h script.
 * DO NOT EDIT THIS FILE.
 * Edit include/makefile.src and the configuration file instead.
 *
 * TODO - actually, with CMake, this needs to turn into a CMake driven
 * configuration file.
 *
 */
#ifndef __RLE_CONFIG_H__
#define __RLE_CONFIG_H__

#define PATCHLEVEL 2

#define USE_STDARG 1
#define USE_PROTOTYPES 1

#define X11 X11
#define ABEKASA60 ABEKASA60
#define ABEKASA62 ABEKASA62
#define ALIAS ALIAS
#define CUBICOMP CUBICOMP
#define CONST_DECL const
#define GIF GIF
#define GRAYFILES GRAYFILES
#define MACPAINT MACPAINT
#define POSTSCRIPT POSTSCRIPT
#define TARGA TARGA
#define TIFF2p4 TIFF2p4
#define VICAR VICAR
#define WASATCH WASATCH
#define WAVEFRONT WAVEFRONT
#define GCC GCC
#define HPUX800CC HPUX800CC
#define NO_RANLIB NO_RANLIB
#define SYS_V_SETPGRP SYS_V_SETPGRP
#define USE_SHARED_LIB USE_SHARED_LIB
#define USE_STRING_H USE_STRING_H
#define USE_TIME_H USE_TIME_H
#define VOID_STAR VOID_STAR
/* -*-C-*- */
/***************** From rle_config.tlr *****************/

/* CONST_DECL must be defined as 'const' or nothing. */
#ifdef CONST_DECL
#undef CONST_DECL
#define CONST_DECL const

#else
#define CONST_DECL

#endif

/* A define for getx11. */
#ifndef USE_XLIBINT_H
#define XLIBINT_H_NOT_AVAILABLE
#endif

/* Typedef for void * so we can use it consistently. */
#ifdef VOID_STAR
typedef	void *void_star;
#else
typedef char *void_star;
#endif

#ifdef USE_STDLIB_H
#include <stdlib.h>
#else

/* Some programs include files from other packages that also declare
 * malloc.  Avoid double declaration by #define NO_DECLARE_MALLOC
 * before including this file.
 */
#ifndef NO_DECLARE_MALLOC
#ifdef USE_PROTOTYPES
#   include <sys/types.h>	/* For size_t. */
    extern void_star malloc( size_t );
    extern void_star calloc( size_t, size_t );
    extern void_star realloc( void_star, size_t );
    extern void free( void_star );
#else
    extern void_star malloc();
    extern void_star realloc();
    extern void_star calloc();
    extern void free();
    extern void cfree();
#endif /* USE_PROTOTYPES */
#endif /* NO_DECLARE_MALLOC */

#ifndef _WIN32
#ifdef USE_PROTOTYPES
extern char *getenv( CONST_DECL char *name );
#else
extern char *getenv();
#endif
#endif

#endif /* USE_STDLIB_H */

#ifdef USE_STRING_H
    /* SYS V string routines. */
#   include <string.h>
#   define index strchr
#   define rindex strrchr
#else
    /* BSD string routines. */
#   include <strings.h>
/* Really, should define USE_STRING_H if __STDC__, but be safe. */
#ifdef __STDC__
#   define index strchr
#   define rindex strrchr
#endif /* __STDC__ */
#endif /* USE_STRING_H */

#ifdef NEED_BSTRING
    /* From bstring.c. */
    /*****************************************************************
     * TAG( bstring bzero )
     * 'Byte string' functions.
     */
#   define bzero( _str, _n )		memset( _str, '\0', _n )
#   define bcopy( _from, _to, _count )	memcpy( _to, _from, _count )
#endif

#ifdef NEED_SETLINEBUF
#   define setlinebuf( _s )	setvbuf( (_s), NULL, _IOLBF, 0 )
#endif

#endif /* __RLE_CONFIG_H__ */

/** @} */


src/tools/rlehdr.c
/*
 * This software is copyrighted as noted below.  It may be freely copied,
 * modified, and redistributed, provided that the copyright notice is
 * preserved on all copies.
 *
 * There is no warranty or other guarantee of fitness for this software,
 * it is provided solely "as is".  Bug reports or fixes may be sent
 * to the author, who may or may not act on them as he desires.
 *
 * You may not include this software in a program or other software product
 * without supplying the source, or without informing the end-user that the
 * source is available for no extra charge.
 *
 * If you modify this software, you should include a notice giving the
 * name of the person performing the modification, the date of modification,
 * and the reason for such modification.
 */
/*
 * rlehdr.c - Print header from an RLE file.
 *
 * Author:	Spencer W. Thomas
 * 		Computer Science Dept.
 * 		University of Utah
 * Date:	Thu Jan 22 1987
 * Copyright (c) 1987, University of Utah
 */
#ifndef lint
char rcsid[] = "$Header$";
#endif
/*
rlehdr()			Tag the file.
*/

#include <stdlib.h>
#include <stdio.h>
#include "rle.h"
#include "rle_config.h"

/* Utah Raster Toolkit major version number. */
#define URT_VERSION 	3.0

void print_hdr(), print_map(), print_codes(), print_brief_hdr();

/*****************************************************************
 * TAG( main )
 *
 * Read and print in human readable form the header of an RLE file.
 *
 * Usage:
 *	rlehdr [-b] [-c comment-name] [-d] [-m] [-v] [files ...]
 * Inputs:
 *  	-b: 	    	Brief.  Fit it on one line.
 *      -c comment-names,...:
    	    	    	The first line of the first comment matching
 *  	    	    	one of these names will be printed.  Sets -b.
 *	-m:		Dump color map contents.
 *	-d:		Dump entire file contents.
 *	-v: 	    	Print URT version number (and ignores files.)
 * 	files:		RLE file(s) to interpret.  If none, reads
 * 	    	    	from standard input.
 * Outputs:
 * 	Prints contents of file header.
 * Assumptions:
 *	[None]
 * Algorithm:
 *	[None]
 */

int
main( argc, argv )
int argc;
char **argv;
{
    CONST_DECL char ** fname = NULL;
    CONST_DECL char *stdname = "-";
    char **comment_names = NULL;
    int     	brief = 0,
		cflag = 0,
    		hflag = 0,
    		num_hdr = 1,
		mflag = 0,
    		suppress = 0,
		dbg_flag = 0,
		nfname = 0,
		version = 0;
    int rle_err = 0, rle_cnt;
    rle_hdr in_hdr;

    in_hdr = *rle_hdr_init( NULL );

    if ( scanargs( argc, argv,
	   "% b%- c%-comment-name!,s h%-n%d Mm%- s%- d%- v%- infile%*s (\n\
\t-b\tBrief mode, one line per image.\n\
\t-c\tWith -b, print 1st line of 1st comment matching a name.\n\
\t-h [n]\tPrint only first (n) header(s) in each file.\n\
\t-m\tPrint contents of colormap.\n\
\t-M\tPrint colormap suitable for rleldmap -t.\n\
\t-s\tSuppress printing header at all (useful with -M).\n\
\t-d\tPrint image data (LOTS of output)\n\
\t-v\tPrint URT version number.  Ignores other arguments.)",
		   &brief, &cflag, &cflag, &comment_names,
		   &hflag, &num_hdr, &mflag, &suppress, &dbg_flag, &version,
		   &nfname, &fname ) == 0 )
	exit( 1 );

    /* If -c was specified, set -b. */
    if ( cflag )
	brief = 1;

    if ( nfname == 0 )
    {
	nfname = 1;
	fname = &stdname;
    }

    if ( version )
    {
	printf( "Utah Raster Toolkit version %3.1f", URT_VERSION );
#if PATCHLEVEL > 0
    	printf( "(patch %d)", PATCHLEVEL );
#endif
	putchar( '\n' );
	exit( 0 );
    }

    for ( ; nfname > 0; fname++, nfname-- )
    {
	in_hdr.rle_file = rle_open_f(cmd_name( argv ), *fname, "r");
	rle_names( &in_hdr, cmd_name( argv ), *fname, 0 );

	for ( rle_cnt = 0;
	      !(hflag && rle_cnt >= num_hdr) &&
	      (rle_err = rle_get_setup( &in_hdr )) == RLE_SUCCESS;
	      rle_cnt++ )
	{
	    /* Separate headers with a newline. */
	    if ( !brief && rle_cnt > 0 )
		putchar( '\n' );

	    if ( !suppress )
		if ( brief )
		    print_brief_hdr( &in_hdr, cflag, comment_names );
		else
		    print_hdr( &in_hdr );
	    if ( mflag )
		print_map( &in_hdr, mflag );
	    if ( dbg_flag )
		print_codes( &in_hdr );
	    else if ( !hflag || rle_cnt + 1 < num_hdr )
		while ( rle_getskip( &in_hdr ) != 32768 )
		    ;
	}

	/* Separate headers with a newline. */
	if ( !brief && nfname > 1 )
	    putchar( '\n' );

	fclose( in_hdr.rle_file );

	/* Check for an error.  EOF or EMPTY is ok if at least one image
	 * has been read.  Otherwise, print an error message.
	 */
	if ( rle_cnt == 0 || (rle_err != RLE_EOF && rle_err != RLE_EMPTY) )
	    rle_get_error( rle_err, cmd_name( argv ), *fname );
    }

    exit( 0 );			/* success */
}


/*****************************************************************
 * TAG( print_hdr )
 *
 * Print the RLE header information given.
 *
 * Inputs:
 *	the_hdr:	Header information.
 * Outputs:
 * 	Prints info on stdout.
 * Assumptions:
 *	[None]
 * Algorithm:
 *	[None]
 */
void
print_hdr( the_hdr )
rle_hdr *the_hdr;
{
    register int i;

    if ( the_hdr->img_num > 1 )
	printf( "RLE header information for %s image %d:\n",
		the_hdr->file_name, the_hdr->img_num );
    else
	printf( "RLE header information for %s:\n", the_hdr->file_name );
    printf( "Originally positioned at (%d, %d), size (%d, %d)\n",
	    the_hdr->xmin, the_hdr->ymin,
	    the_hdr->xmax - the_hdr->xmin + 1,
	    the_hdr->ymax - the_hdr->ymin + 1 );
    printf( "Saved %d color channels%s\n", the_hdr->ncolors,
	    the_hdr->alpha ? " plus Alpha channel" : "" );
    if ( the_hdr->background == 0 )
	printf( "No background information was saved\n" );
    else
    {
	if ( the_hdr->background == 1 )
	    printf( "Saved in overlay mode with original background color" );
	else
	    printf( "Screen will be cleared to background color" );
	for ( i = 0; i < the_hdr->ncolors; i++ )
	    printf( " %d", the_hdr->bg_color[i] );
	putchar( '\n' );
    }

    if ( the_hdr->ncmap > 0 )
	printf( "%d channels of color map %d entries long were saved.\n",
		the_hdr->ncmap, 1 << the_hdr->cmaplen );

    if ( the_hdr->comments != NULL && *the_hdr->comments != NULL )
    {
	printf( "Comments:\n" );
	for ( i = 0; the_hdr->comments[i] != NULL; i++ )
	    printf( "%s\n", the_hdr->comments[i] );
    }

    fflush( stdout );
}

/*****************************************************************
 * TAG( print_brief_hdr )
 *
 * Print the RLE header information on one line.
 *
 * Inputs:
 *	the_hdr:	Header information.
 *	ncomment:   	Number of comment names to test.
 *	comment_names:	Print the first line of the first matching
 *	    	    	comment found.
 * Outputs:
 * 	Prints info on stdout.
 * Assumptions:
 *	[None]
 * Algorithm:
 *	[None]
 */
void
print_brief_hdr( the_hdr, ncomment, comment_names )
rle_hdr *the_hdr;
int ncomment;
char **comment_names;
{
    register int i;

    if ( the_hdr->img_num > 1 )
	printf( "%s(%d):",
		the_hdr->file_name, the_hdr->img_num );
    else
	printf( "%s:", the_hdr->file_name );
    printf( "\t[%d,%d]+[%d,%d]",
	    the_hdr->xmin, the_hdr->ymin,
	    the_hdr->xmax - the_hdr->xmin + 1,
	    the_hdr->ymax - the_hdr->ymin + 1 );
    printf( "x%d%s", the_hdr->ncolors,
	    the_hdr->alpha ? "+A" : "" );
    if ( the_hdr->background == 0 )
	;
    else
    {
	if ( the_hdr->background == 1 )
	    printf( ", OV=" );
	else
	    printf( ", BG=" );
	for ( i = 0; i < the_hdr->ncolors; i++ )
	    printf( "%d%s", the_hdr->bg_color[i],
		    i == the_hdr->ncolors - 1 ? " " : "," );
    }

    if ( the_hdr->ncmap > 0 )
	printf( ", map %dx%d",
		the_hdr->ncmap, 1 << the_hdr->cmaplen );

    if ( the_hdr->comments != NULL && *the_hdr->comments != NULL )
    {
	char *the_comment, *cp;

	if ( ncomment > 0 )
	{
	    for ( ; ncomment > 0; ncomment--, comment_names++ )
		if (the_comment = rle_getcom( *comment_names, the_hdr ))
		{
		    if ( (cp = index( the_comment, '\n' )) )
			printf( ", %s=%.*s", *comment_names,
				cp - the_comment - 1, the_comment );
		    else
			printf( ", %s=%s", *comment_names, the_comment );
		    break;
		}
	}
	if ( ncomment == 0 )
	    printf( ", (C)" );
    }

    putchar( '\n' );

    fflush( stdout );
}

/*****************************************************************
 * TAG( print_map )
 *
 * Print the color map from a the_hdr structure.
 * Inputs:
 * 	the_hdr:	Sv_hdr structure containing color map.
 * 	mflag:		If 1, print "human readable".  If 2, print in
 * 			form suitable for rleldmap -t.
 * Outputs:
 * 	Prints color map (if present) on stdout.
 * Assumptions:
 *	color_map_length comment is relevant.
 * Algorithm:
 *	[None]
 */
void
print_map( the_hdr, mflag )
rle_hdr *the_hdr;
int mflag;
{
    register int i, j;
    int c, maplen, ncmap, nmap;
    rle_map * cmap;
    char *len_com;

    if ( the_hdr->ncmap == 0 )
	return;

    maplen = (1 << the_hdr->cmaplen);
    ncmap = the_hdr->ncmap;
    cmap = the_hdr->cmap;

    if ( (len_com = rle_getcom( "color_map_length", the_hdr )) != NULL &&
	 atoi( len_com ) > 0 )
	nmap = atoi( len_com );
    else
	nmap = maplen;

    if ( mflag == 1 )
    {
	printf( "Color map contents, values are 16-bit(8-bit):\n" );
	for ( i = 0; i < nmap; i++ )
	{
	    printf( "%3d:\t", i );
	    for ( j = 0, c = 0; j < ncmap; j++, c += maplen )
		printf( "%5d(%3d)%c", cmap[i+c], cmap[i+c] >> 8,
			j == ncmap - 1 ? '\n' : '\t' );

	}
    }
    else
	for ( i = 0; i < nmap; i++ )
	{
	    for ( j = 0, c = 0; j < ncmap; j++, c += maplen )
		printf( "%3d%c", cmap[i+c] >> 8,
			j == ncmap - 1 ? '\n' : '\t' );

	}
}


/*****************************************************************
 * TAG( print_codes )
 *
 * Print the RLE opcodes in an RLE file.
 * Inputs:
 * 	the_hdr:		Header for RLE file (already open).
 * Outputs:
 * 	Prints file contents on stderr.
 * Assumptions:
 *
 * Algorithm:
 *	[None]
 */
void
print_codes( the_hdr )
rle_hdr *the_hdr;
{
    rle_pixel ** scans;

    if ( rle_row_alloc( the_hdr, &scans ) < 0 )
	RLE_CHECK_ALLOC( "rlehdr", 0, 0 );

    rle_debug(1);
    while ( !feof( the_hdr->rle_file ) &&
	    !the_hdr->priv.get.is_eof )
	rle_getrow( the_hdr, scans );

    rle_row_free( the_hdr, scans );
}
